rts/lra/jaxrs/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java

```diff
@@ -460,46 +575,149 @@ private boolean isJaxRsCancel(ContainerRequestContext requestContext, ContainerR
+        }
     }
 
-    private String getCompensatorId(URI lraId, UriInfo uriInfo, Long timeout) {
-        Map<String, String> terminateURIs = NarayanaLRAClient.getTerminationUris(resourceInfo.getResourceClass(), uriInfo, timeout);
+    // convert the list of steps carried out by the filters into a warning message
+    private String processLRAOperationFailures(ArrayList<Progress> progress) {
+        StringJoiner badOps = new StringJoiner(", ");
+        StringJoiner goodOps = new StringJoiner(", ");
+        StringBuffer code = new StringBuffer("-");
 
-        if (!terminateURIs.containsKey("Link")) {
-            throwGenericLRAException(lraId, Response.Status.BAD_REQUEST.getStatusCode(),
-                    "Missing complete or compensate annotations");
-        }
+        progress.forEach(p -> {
+            if (p.ok()) {
+                code.insert(0, p.progress.ordinal());
+                goodOps.add(String.format("%s (%s)", p.progress.name(), p.progress.status));
+            } else {
+                code.append(p.progress.ordinal());
+                badOps.add(String.format("%s (%s)", p.progress.name(), p.reason));
+            }
+        });
+
+        /*
+         * return a string which encodes the result:
+         * <major code>-<failed op codes>-<successful op codes>: <details of failed ops> (<details of successful ops>)
+         *
+         * where
+         *
+         * <major code>: corresponds to the id of the message in the logs
+         * <failed op codes>: each digit corresponds to the enum ordinal calue of the ProgressStep enum value that was successful
+         * <successful op codes>: each digit corresponds to the enum ordinal calue of the ProgressStep enum value that failed
+         * <details of failed ops>: comma separated list of failed operation details "<op name> (<exception message>)"
+         * <details of successful ops>: comma separated list of successful operation details "<op name> (<op description>)"
+         */
+
+        return badOps.length() != 0 ? String.format("%d-%s: %s (%s)", i18nMessageCode, code, badOps, goodOps) : null;
+    }
 
-        return terminateURIs.get("Link");
+    private boolean progressDoesNotContain(ArrayList<Progress> progress, ProgressStep step) {
+        return progress.stream().noneMatch(p -> p.progress == step);
     }
 
-    private void lraTrace(URI lraId, String reason) {
-        if (LRALogger.logger.isTraceEnabled()) {
-            Method method = resourceInfo.getResourceMethod();
-            LRALogger.logger.tracef("%s: container request for method %s: lra: %s%n",
-                    reason, method.getDeclaringClass().getName() + "#" + method.getName(),
-                    lraId == null ? "context" : lraId);
+    // add another step to the list of steps performed so far
+    private ArrayList<Progress> updateProgress(ArrayList<Progress> progress, ProgressStep step, String reason) {
+        if (progress == null) {
```

Comment:
Could this be called from multiple threads?

Reply:
I missed these comments, github had collapse them all. Anyway,

This code is part of a request response processing filter chain. The context for such a request is passed into the filter method and applies to the particular request context. If you check the variable `progress` you will see that it is a method local variable and therefore not subject to your implied concurrency concern. In addition the contents of `progress` are passed to the response filter via a context property (of the request context)n and again there are no concurrency concerns.